sql: support prepared for SHOW

madelynnblue opened the pull request:
Although psql errors on `prepare a as SHOW TIME ZONE`, postgres does
indeed support prepared SHOW statements. The psql client doesn't support
the extended prepare protocol, so it passes on the parsing of those
to the server, which errors. This was why I thought we didn't need to
support prepared SHOW. However using a better driver indicates we need
to support this.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4150)

<!-- Reviewable:end -->

tamird commented on sql/plan.go:
```diff
@@ -165,6 +165,18 @@ func (p *planner) prepare(stmt parser.Statement) (planNode, *roachpb.Error) {
+	case *parser.Show:
+		return p.Show(n)
+	case *parser.ShowColumns:
+		return p.ShowColumns(n)
```
I think each of these deserves a test.

madelynnblue committed:
sql: support prepared for SHOW

Although psql errors on `prepare a as SHOW TIME ZONE`, postgres does
indeed support prepared SHOW statements. The psql client doesn't support
the extended prepare protocol, so it passes on the parsing of those
to the server, which errors. This was why I thought we didn't need to
support prepared SHOW. However using a better driver indicates we need
to support this.

mjibson commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/plan.go, line 171 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4150#-K9i0ocuODguHphJQbcO-r1-171:-K9iDFUbagwkQbPfi1EZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5a17b7e9d4cc80ac7c919055dc071cf5b0937ad/sql/plan.go#L171)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4150)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4150)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4150#-:-K9iDbq2DFvVqbPA8EPY:676834233)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

Is it possible postgres supports (protocol) prepared statements for all statement types? I recall (from the grammar) the `PREPARE` syntax only allows `{INSERT,UPDATE,DELETE,SELECT}`. But prepared statements in the protocol are different.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4150#-:-K9iwoYAZC53zkPIGQcD:997521325)*
<!-- Sent from Reviewable.io -->

mjibson commented:
That is my suspicion (that the protocol supports prepare for all statement types). I will pursue this more next week.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4150#-:-K9kK7UkDPtxgIRQStaM:-1359066270)*
<!-- Sent from Reviewable.io -->
